File: src/lib/api/git/branches.ts
// Branch lifecycle and the history grafts that run against a checked-out
// branch: checkout, create, delete, rename, upstream, merge, fast-forward,
// rebase, reset, cherry-pick, revert. Mirrors `commands/branches.rs`.

import { invoke } from "@/lib/api/invoke";
import { parse } from "@/lib/api/validate";
import { z } from "zod";

export const branchesApi = {
  checkout: async (path: string, target: string, detached = false) =>
    parse(z.string(), await invoke("checkout", { path, target, detached }), "checkout"),

  /** Check out the local counterpart of `remote/branch`, creating it with
   * tracking or safely fast-forwarding the existing local branch. */
  checkoutRemoteBranch: async (path: string, remote: string, branch: string) =>
    parse(
      z.string(),
      await invoke("checkout_remote_branch", { path, remote, branch }),
      "checkout_remote_branch",
    ),

  /** Create a branch at `startPoint` (the ref the user picked, so a
   * remote-tracking start point keeps git's automatic upstream setup), pinned
   * to the `expectedOid` the user saw. */
  createBranch: async (path: string, name: string, startPoint: string, expectedOid: string) =>
    parse(
      z.string(),
      await invoke("create_branch", { path, name, startPoint, expectedOid }),
      "create_branch",
    ),

  deleteBranch: async (path: string, name: string, expectedOid: string, force = false) =>
    parse(
      z.string(),
      await invoke("delete_branch", { path, name, expectedOid, force }),
      "delete_branch",
    ),

  renameBranch: async (path: string, oldName: string, newName: string) =>
    parse(
      z.string(),
      await invoke("rename_branch", { path, old: oldName, new: newName }),
      "rename_branch",
    ),

  /** Point `branch`'s upstream at the remote-tracking ref `upstream` (e.g.
   * "origin/main"). The ref must already exist. */
  setUpstream: async (path: string, branch: string, upstream: string) =>
    parse(z.string(), await invoke("set_upstream", { path, branch, upstream }), "set_upstream"),

  mergeBranch: async (
    path: string,
    source: string,
    expectedSourceOid: string,
    destination: string | null,
    expectedDestinationOid: string,
  ) =>
    parse(
      z.string(),
      await invoke("merge_branch", {
        path,
        source,
        expectedSourceOid,
        destination,
        expectedDestinationOid,
      }),
      "merge_branch",
    ),

  /** Fast-forward the explicit branch from its expected oid to a captured
   * target oid. Rust chooses the live checked-out/non-checked-out mechanism. */
  fastForwardBranch: async (
    path: string,
    branch: string,
    expectedBranchOid: string,
    targetOid: string,
  ) =>
    parse(
      z.string(),
      await invoke("fast_forward_branch", { path, branch, expectedBranchOid, targetOid }),
      "fast_forward_branch",
    ),

  /** Rebase the explicit `source` branch/revision onto `onto`; the backend
   * carries both operands through one git process instead of trusting HEAD. */
  rebaseOnto: async (path: string, source: string, expectedSourceOid: string, ontoOid: string) =>
    parse(
      z.string(),
      await invoke("rebase_onto", { path, source, expectedSourceOid, ontoOid }),
      "rebase_onto",
    ),

  resetTo: async (
    path: string,
    source: string | null,
    expectedSourceOid: string | null,
    targetOid: string,
    mode: "soft" | "mixed" | "hard",
    expectedState?: string | null,
    expectedHeadBranch?: string | null,
    expectedHeadOid?: string | null,
  ) =>
    parse(
      z.string(),
      await invoke("reset_to", {
        path,
        source,
        expectedSourceOid,
        targetOid,
        mode,
        expectedState: expectedState ?? null,
        expectedHeadBranch: expectedHeadBranch ?? null,
        expectedHeadOid: expectedHeadOid ?? null,
      }),
      "reset_to",
    ),

  /** Cherry-pick several commits in one atomic `git cherry-pick A B C…`. */
  cherryPickMany: async (
    path: string,
    expectedBranch: string | null,
    expectedOid: string,
    commits: string[],
  ) =>
    parse(
      z.string(),
      await invoke("cherry_pick_many", { path, expectedBranch, expectedOid, commits }),
      "cherry_pick_many",
    ),

  /** Revert several commits in one atomic `git revert --no-edit A B…`. */
  revertMany: async (
    path: string,
    expectedBranch: string | null,
    expectedOid: string,
    commits: string[],
  ) =>
    parse(
      z.string(),
      await invoke("revert_many", { path, expectedBranch, expectedOid, commits }),
      "revert_many",
    ),
};

File: src-tauri/src/commands/branches.rs
//! Branch checkout/create/delete/rename and history-rewriting operations (merge, rebase, reset, cherry-pick, revert).

use super::blocking;
use crate::git;

#[tauri::command]
pub async fn checkout(path: String, target: String, detached: bool) -> Result<String, String> {
    blocking(move || git::write::branch_checkout::checkout(&path, &target, detached)).await
}

#[tauri::command]
pub async fn checkout_remote_branch(
    path: String,
    remote: String,
    branch: String,
) -> Result<String, String> {
    blocking(move || git::write::branch_checkout::checkout_remote_branch(&path, &remote, &branch))
        .await
}

#[tauri::command]
pub async fn create_branch(
    path: String,
    name: String,
    start_point: String,
    expected_oid: String,
) -> Result<String, String> {
    blocking(move || git::write::branches::create_branch(&path, &name, &start_point, &expected_oid))
        .await
}

#[tauri::command]
pub async fn delete_branch(
    path: String,
    name: String,
    expected_oid: String,
    force: bool,
) -> Result<String, String> {
    blocking(move || git::write::branches::delete_branch(&path, &name, &expected_oid, force)).await
}

#[tauri::command]
pub async fn rename_branch(path: String, old: String, new: String) -> Result<String, String> {
    blocking(move || git::write::branches::rename_branch(&path, &old, &new)).await
}

#[tauri::command]
pub async fn set_upstream(
    path: String,
    branch: String,
    upstream: String,
) -> Result<String, String> {
    blocking(move || git::write::branches::set_upstream(&path, &branch, &upstream)).await
}

#[tauri::command]
pub async fn merge_branch(
    path: String,
    source: String,
    expected_source_oid: String,
    destination: Option<String>,
    expected_destination_oid: String,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::merge_into(
            &path,
            &source,
            &expected_source_oid,
            destination.as_deref(),
            &expected_destination_oid,
        )
    })
    .await
}

#[tauri::command]
pub async fn fast_forward_branch(
    path: String,
    branch: String,
    expected_branch_oid: String,
    target_oid: String,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::fast_forward_branch_at(
            &path,
            &branch,
            &expected_branch_oid,
            &target_oid,
        )
    })
    .await
}

#[tauri::command]
pub async fn rebase_onto(
    path: String,
    source: String,
    expected_source_oid: String,
    onto_oid: String,
) -> Result<String, String> {
    blocking(move || git::write::history::rebase(&path, &source, &expected_source_oid, &onto_oid))
        .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)] // Tauri command shape mirrors the frontend IPC contract.
pub async fn reset_to(
    path: String,
    source: Option<String>,
    expected_source_oid: Option<String>,
    target_oid: String,
    mode: String,
    expected_state: Option<String>,
    expected_head_branch: Option<String>,
    expected_head_oid: Option<String>,
) -> Result<String, String> {
    let request = git::write::reset::ResetRequest::parse(
        source.as_deref(),
        expected_source_oid.as_deref(),
        &mode,
        expected_state.as_deref(),
        expected_head_branch.as_deref(),
        expected_head_oid.as_deref(),
    )?;
    blocking(move || git::write::reset::reset_branch(&path, &target_oid, request)).await
}

#[tauri::command]
pub async fn cherry_pick(
    path: String,
    expected_branch: Option<String>,
    expected_oid: String,
    commit: String,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::cherry_pick_onto(
            &path,
            expected_branch.as_deref(),
            &expected_oid,
            &commit,
        )
    })
    .await
}

#[tauri::command]
pub async fn cherry_pick_many(
    path: String,
    expected_branch: Option<String>,
    expected_oid: String,
    commits: Vec<String>,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::cherry_pick_many_onto(
            &path,
            expected_branch.as_deref(),
            &expected_oid,
            &commits,
        )
    })
    .await
}

#[tauri::command]
pub async fn revert_commit(
    path: String,
    expected_branch: Option<String>,
    expected_oid: String,
    commit: String,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::revert_onto(&path, expected_branch.as_deref(), &expected_oid, &commit)
    })
    .await
}

#[tauri::command]
pub async fn revert_many(
    path: String,
    expected_branch: Option<String>,
    expected_oid: String,
    commits: Vec<String>,
) -> Result<String, String> {
    blocking(move || {
        git::write::history::revert_many_onto(
            &path,
            expected_branch.as_deref(),
            &expected_oid,
            &commits,
        )
    })
    .await
}
